core/server/proxy/src/main/java/alluxio/proxy/s3/TaggingData.java

```diff
@@ -111,12 +110,69 @@ public String toString() {
     return mTagSet.toString();
   }
 
+  /**
+   * Returns the Map object containing the contents of
+   * the TagSet. Useful for parsing or iterating over tag data.
+   * @return a Map object containing the tags
+   */
+  @XmlTransient
+  public Map<String, String> getTagMap() {
+    return mTagMap;
+  }
+
+  /**
+   * Adds the provided key-value pair as a new tag
+   * Overwrites any duplicate tags.
+   * @param key
+   * @param value
+   * @return a reference to this object
+   */
+  @XmlTransient
+  private void addTag(String key, String value) {
+    if (!mTagMap.containsKey(key)) {
+      mTagSet.mTags.add(new TagObject(key, value));
+      mTagMap.put(key, value);
+    } // else, we have to update the existing tag
+    for (TagObject tag : mTagSet.mTags) {
+      if (tag.mKey.equals(key)) {
+        tag.mValue = value;
+        break;
+      }
+    }
+    mTagMap.put(key, value);
+  }
+
+  /**
+   * Merges the contents of the provided tag map with
+   * the existing tags. Overwrites any duplicate tags.
+   * @param tagMap
+   * @return a reference to this object
+   */
+  @XmlTransient
+  public TaggingData addTags(Map<String, String> tagMap) {
+    for (Map.Entry<String, String> tag : tagMap.entrySet()) {
+      addTag(tag.getKey(), tag.getValue());
+    }
+    validateTags();
+    return this;
+  }
+
+  /**
+   * Fills the tag map with the contents of the tag set.
+   */
+  @XmlTransient
+  private void repopulateTagMap() {
+    mTagMap.clear();
+    for (TagObject tag : mTagSet.mTags) {
+      mTagMap.put(tag.mKey, tag.mValue);
+    }
+  }
+
```

Comment:
I added a separate data structure to mirror the underlying XML structure for convenience when trying to iterate on/modify said data.